src/backend/session/libseat.rs

```diff
@@ -0,0 +1,279 @@
+                .map_err(|err| Error::FailedToChangeVt(Errno::from_i32(err.into())))
+        } else {
+            Err(Error::SessionLost)
+        }
+    }
+
+    fn is_active(&self) -> bool {
+        if let Some(internal) = self.internal.upgrade() {
+            internal.active.load(Ordering::SeqCst)
+        } else {
+            false
+        }
+    }
+
+    fn seat(&self) -> String {
+        self.seat_name.clone()
+    }
+}
+
+impl LibSeatSessionNotifier {
+    /// Creates a new session object belonging to this notifier.
+    pub fn session(&self) -> LibSeatSession {
+        LibSeatSession {
+            internal: Rc::downgrade(&self.internal),
+            seat_name: self.internal.seat.borrow_mut().name().to_owned(),
+        }
+    }
+
+    /// Get a handle to the Signaler of this session.
+    ///
+    /// You can use it to listen for signals generated by the session.
+    pub fn signaler(&self) -> Signaler<SessionSignal> {
+        self.signaler.clone()
+    }
+}
+
+impl EventSource for LibSeatSessionNotifier {
+    type Event = ();
+    type Metadata = ();
+    type Ret = ();
+
+    fn process_events<F>(&mut self, _readiness: Readiness, _token: Token, _: F) -> std::io::Result<()>
+    where
+        F: FnMut((), &mut ()),
+    {
+        self.internal.seat.borrow_mut().dispatch(0).unwrap();
+        Ok(())
+    }
+
+    fn register(&mut self, poll: &mut Poll, token: Token) -> std::io::Result<()> {
+        poll.register(
+            self.internal.seat.borrow_mut().get_fd().unwrap(),
+            calloop::Interest::READ,
+            calloop::Mode::Level,
+            token,
+        )
+        .unwrap();
+        Ok(())
+    }
+
+    fn reregister(&mut self, poll: &mut Poll, token: Token) -> std::io::Result<()> {
+        poll.reregister(
+            self.internal.seat.borrow_mut().get_fd().unwrap(),
+            calloop::Interest::READ,
+            calloop::Mode::Level,
+            token,
+        )
+        .unwrap();
+        Ok(())
+    }
+
+    fn unregister(&mut self, poll: &mut Poll) -> std::io::Result<()> {
+        poll.unregister(self.internal.seat.borrow_mut().get_fd().unwrap())
+            .unwrap();
+        Ok(())
+    }
+}
+
+/// Errors related to direct/tty sessions
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    /// Failed to open session
+    #[error("Failed to open session: {0}")]
+    FailedToOpenSession(Errno),
+
+    /// Failed to open device
+    #[error("Failed to open device: {0}")]
+    FailedToOpenDevice(Errno),
+
+    /// Failed to close device
+    #[error("Failed to close device: {0}")]
+    FailedToCloseDevice(Errno),
+
+    /// Failed to close device
+    #[error("Failed to change vt: {0}")]
+    FailedToChangeVt(Errno),
+
+    /// Session is already closed,
+    #[error("Session is already closed")]
+    SessionLost,
+
+    /// Unknown
+    #[error("Unknown")]
+    Unknown,
+}
+
+impl AsErrno for Error {
+    fn as_errno(&self) -> Option<i32> {
+        None
+    }
+}
```

Comment:
What is this impl used for?

Reply:
As far as I know, it is not used anywhere, I simply saw that dbus, direct and auto backends all have one, so I added one here as well, I just forgot to return proper errnos with it.

It's worth noting that both dbus and auto always return None when `as_errno` is called, so auto will never forward libseat errnos to caller

Reply:
I think that was previously used to convert most session errors into `std::io::Error`s or `nix::Error`s. This got possibly removed when we switched from `error-chain` to `thiserror`.